### PR TITLE
Added specifications for refactoring tendrl code

### PR DESCRIPTION
**Pull request by shtripat**
This specification details the refactoring across the tendrl
modules to move major common code to common module.

Signed-off-by: Shubhendu <shtripat@redhat.com>

### CODE REVIEW

**Comment by shtripat** [comment]
@nthomas-redhat @r0h4n @brainfunked @anmolbabu @nnDarshan @GowthamShanmugam please review.

**Review by r0h4n — CHANGES_REQUESTED** [review]
Separate out refactoring of each component into individual specs, common, tendrl-gluster-integraton etc

**Comment by shtripat** [comment]
@r0h4n I was of a thought that each module specific code changes would be created as individual github issues and this spec PR would be linked to all of them. So effectively this spec PR ends up into multiple github issues and multiple re-factoring code change PRs as a result.

@nthomas-redhat @brainfunked comments?? ^^

If everybody strongly feels, its required, I can create multiple ones (its just a copy paste of same content mostly ;))

File: specs/refactor_common_code_and_move.adoc [hunk 1]
```diff
@@ -0,0 +1,206 @@
+= Refactor modules like tendrl-gluster-integartion, tendrl-node-agent etc and
```
Comment (brainfunked) [inline specs/refactor_common_code_and_move.adoc]: Line breaks do not work in titles, if I'm not mistaken. Shorten the main title. Add a subtitle if you wish to convey additional context.

File: specs/refactor_common_code_and_move.adoc [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+business logic.
+
+Even there are various sample configuration files which are replicated across
+the tendrl modules and could be more organized and moved to common module. Only
+modules specific configurations should be kept as part of the individual modules
+
+
+== Use Cases
+
+None
```
Comment (brainfunked) [inline specs/refactor_common_code_and_move.adoc]: The most important use case is that this enables commons to function as a library and the code becomes completely reusable by any component.

**Review by brainfunked — CHANGES_REQUESTED** [review]
Need implementation details regarding which specific flows need to be moved.

**Comment by brainfunked** [comment]
The following issues need to be incorporated in this specification:
* https://github.com/Tendrl/common/issues/58
* https://github.com/Tendrl/common/issues/60
* https://github.com/Tendrl/common/issues/62

**Comment by r0h4n** [comment]
@shtripat Specs needs to be per component and github issues should be against that component, It is difficult to track one spec spanning multiple components. Also that way component owners can review those specs without needing much insight on other components apart from making sure the impact on other components is documented

**Comment by brainfunked** [comment]
@r0h4n Specifications are tendrl-global. They can affect multiple components. The specification should explicitly point out the affected components and the changes required across them. Issues from the specification can be created across multiple components. The issues detail out the exact implementation changes required in each of the components.

Basically, specifications are our tendrl project design. Issues are per component implementation details.

File: specs/refactor_common_code_and_move.adoc (hunk 1 above)
Comment (shtripat) [reply specs/refactor_common_code_and_move.adoc]: Ok. Will shorten the heading..

File: specs/refactor_common_code_and_move.adoc (hunk 2 above)
Comment (shtripat) [reply specs/refactor_common_code_and_move.adoc]: Done. Would add the use case details.

**Comment by shtripat** [comment]
@brainfunked Agree. Even I feel this specification could be single and we can add all the required details here and then link the github issues to this from different modules of tendrl.
I would prefer retaining this as single spec.

**Comment by nnDarshan** [comment]
@shtripat Can you please mention about moving the general purpose utilities like command executor, package installer, service manager etc  to common in this spec.
I see you have already mentioned about utils, but they are more specific to flows, manager etc.

**Comment by shtripat** [comment]
@nnDarshan regarding your comment, I think enough details are available under "Proposed changes" and "Work Items". These section talk about manager, persister changes etc in detail. Anything specific you need?

**Comment by nnDarshan** [comment]
@shtripat Can you please mention about general purpose utilities like command executor, package installer, service manager.

File: specs/refactor_common_code_and_move.adoc [hunk 3]
```diff
@@ -203,4 +232,8 @@ None
 
 == References
 
-None
+* https://github.com/Tendrl/common/issues/58
+
+* https://github.com/Tendrl/common/issues/60
+
+* https://github.com/Tendrl/common/issues/62
```
Comment (cybernth) [inline specs/refactor_common_code_and_move.adoc:257]: Have a mention about issue 60 as well in the document. That means common module specific logging and configuirations will be removed. This has a documentaion impact as well

Also you need to create individual issues per component(changes occuring) and link it here

**Review by cybernth — CHANGES_REQUESTED** [review]
overall LGTM if the comments are addressed

File: specs/refactor_common_code_and_move.adoc (hunk 3 above)
Comment (shtripat) [reply specs/refactor_common_code_and_move.adoc:257]: Done

Comment (nnDarshan) [reply specs/refactor_common_code_and_move.adoc:257]: @shtripat can you please include: https://github.com/Tendrl/common/issues/80. Sorry for informing you late.

Comment (shtripat) [reply specs/refactor_common_code_and_move.adoc:257]: Done

**Review by cybernth — APPROVED** [review]
Looks good to me

**Review by r0h4n — APPROVED** [review]

**Pull request merged**